### PR TITLE
CTCP-5975 Add Draft No Longer Available page and functionality

### PR DESCRIPTION
**Pull request by karahanozturk**
https://jira.tools.tax.service.gov.uk/browse/CTCP-5975

image

### CODE REVIEW

File: app/controllers/DraftNoLongerAvailableController.scala [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+class DraftNoLongerAvailableController @Inject() (
+  override val messagesApi: MessagesApi,
+  val controllerComponents: MessagesControllerComponents,
+  view: DraftNoLongerAvailableView
+) extends FrontendBaseController
+    with I18nSupport {
+
+  def onPageLoad(): Action[AnyContent] = Action {
+    implicit request =>
+      NotAcceptable(view())
```
Comment (jackheslop96) [inline app/controllers/DraftNoLongerAvailableController.scala:35]: The API to return 4xx (more investigation needed to see if 406 would be "correct" in this case) but the frontend can probably just return Ok(view)

File: app/controllers/actions/CrossoverAction.scala [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+class CrossoverAction(config: PhaseConfig)(implicit val executionContext: ExecutionContext) extends ActionFilter[DataRequest] with Logging {
+
+  private val TransitionApiVersion = "2.0"
+  private val isConfigTransitional = config.values.apiVersion.toString == TransitionApiVersion
+
+  override protected def filter[A](request: DataRequest[A]): Future[Option[Result]] =
+    Future.successful {
+      if isConfigTransitional != request.userAnswers.isTransitional then Some(Redirect(controllers.routes.DraftNoLongerAvailableController.onPageLoad()))
```
Comment (jackheslop96) [inline app/controllers/actions/CrossoverAction.scala:42]: This action could possibly be obsolete. What we could do is add logic to the data retrieval action to match on the 4xx (406?) response and redirect to DraftNoLongerAvailableController

File: app/models/UserAnswers.scala [hunk 3]
```diff
@@ -29,7 +29,8 @@ final case class UserAnswers(
   data: JsObject = Json.obj(),
   tasks: Map[String, TaskStatus] = Map(),
   status: SubmissionState.Value,
-  departureId: Option[String] = None
+  departureId: Option[String] = None,
+  isTransitional: Boolean = false
```
Comment (jackheslop96) [inline app/models/UserAnswers.scala:33]: We then wouldn't need to add isTransitional to our frontend user answers if it is only used in the API

File: app/views/DraftNoLongerAvailableView.scala.html [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    <p class="govuk-body">
+        <a class="govuk-link" id="make-new-departure" href="@routes.RedirectController.onPageLoad()">@messages("draftNoLongerAvailable.newDeparture")</a>
+    </p>
+
+    <div>
+        @button(
+            messageKey = messages("draftNoLongerAvailable.button"),
+            href = Some(config.manageTransitMovementsDraftDeparturesUrl)
+        )
```
Comment (jackheslop96) [inline app/views/DraftNoLongerAvailableView.scala.html:46]: This was flagged up in our most recent DAC - essentially this is a link disguised as a button. What we should look to do here instead is add a POST method to the controller which does the redirect for us. See example:
- https://github.com/hmrc/manage-transit-movements-unloading-frontend/blob/1c244cbd2a44b1f1c17fc72f1ecc637b4f07ec04/app/views/UnloadingFindingsView.scala.html#L44-L46
- https://github.com/hmrc/manage-transit-movements-unloading-frontend/blob/1c244cbd2a44b1f1c17fc72f1ecc637b4f07ec04/app/controllers/UnloadingFindingsController.scala#L47-L52
-

File: app/controllers/actions/CrossoverAction.scala (hunk 2 above)
Comment (jackheslop96) [reply app/controllers/actions/CrossoverAction.scala:42]: You could add a recover case to the CacheConnector get method

Maybe define a new response model trait with a http reads that covers

- 200 -> Some(user answers)
- 404 -> None
- 406 -> Some case object that indicates the answers exist but are no longer available
- (anything else would be a genuine exception)

Similar to the [GroupEnrolmentResponse](https://github.com/hmrc/manage-transit-movements-departure-frontend/blob/main/app/models/GroupEnrolmentResponse.scala) model.

File: conf/app.routes [hunk 5]
```diff
@@ -3,6 +3,8 @@
 ->         /                                              app.preTaskList.Routes
 
 GET        /drafts/:lrn                                   controllers.DraftController.draftRedirect(lrn: models.LocalReferenceNumber)
+POST       /drafts/redirect-declarations                  controllers.DraftController.onSubmit()
```
Comment (jackheslop96) [inline conf/app.routes:6]: POST url can be the same as the new GET url

File: app/controllers/DraftController.scala [hunk 6]
```diff
@@ -42,4 +42,11 @@ class DraftController @Inject() (
           Redirect(controllers.routes.TaskListController.onPageLoad(lrn))
       }
   }
+
+  def onSubmit(): Action[AnyContent] = Action {
+    Redirect(
+      frontendAppConfig.manageTransitMovementsDraftDeparturesUrl
+    )
+  }
```
Comment (jackheslop96) [inline app/controllers/DraftController.scala:50]: This can move to DraftNoLongerAvailableController

File: app/models/UserAnswers.scala [hunk 7]
```diff
@@ -30,7 +30,7 @@ final case class UserAnswers(
   tasks: Map[String, TaskStatus] = Map(),
   status: SubmissionState.Value,
   departureId: Option[String] = None
-) {
+) extends UserAnswersResponse {
```
Comment (jackheslop96) [inline app/models/UserAnswers.scala:33]: Ideally UserAnswersResponse would be a sealed trait, and we'd have some case class containing the user answers

File: app/models/UserAnswersResponse.scala [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+ * limitations under the License.
+ */
+
+package models
+
+import play.api.http.Status.*
+import play.api.libs.json.{Json, Reads}
+import uk.gov.hmrc.http.{HttpReads, HttpResponse}
+
+trait UserAnswersResponse
```
Comment (jackheslop96) [inline app/models/UserAnswersResponse.scala:23]: So something like

```
sealed trait UserAnswersResponse

object UserAnswersResponse {

  case class Answers(value: UserAnswers) extends UserAnswersResponse

  case object NoAnswers extends UserAnswersResponse

  case object NotAcceptable extends UserAnswersResponse

  case class Other(code: Int, message: String) extends UserAnswersResponse

  implicit val httpReads: HttpReads[UserAnswersResponse] =
    (_: String, _: String, response: HttpResponse) =>
      response.status match {
        case OK =>
          response.json
            .validate[UserAnswers]
            .map(Answers.apply)
            .fold(
              errors => Other(response.status, s"Failed to validate json: $errors"),
              identity
            )
        case NOT_FOUND      => NoAnswers
        case NOT_ACCEPTABLE => NotAcceptable
        case _              => Other(response.status, response.body)
      }
}
```

File: app/models/UserAnswersResponse.scala [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+object UserAnswersResponse {
+
+  case object NoAnswers extends UserAnswersResponse
+
+  case class BadRequest(code: String) extends UserAnswersResponse
+
+  object BadRequest {
+
+    implicit val reads: Reads[BadRequest] = Json.reads[BadRequest]
+  }
```
Comment (jackheslop96) [inline app/models/UserAnswersResponse.scala:34]: The API doesn't return a bad request, so this is redundant

File: app/models/UserAnswersResponse.scala [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        response.json
+          .validate[T]
+          .fold(
+            errors => Other(response.status, s"Failed to validate json: $errors"),
+            identity
+          )
+
+      response.status match {
+        case OK                     => validate[UserAnswers]
+        case NO_CONTENT | NOT_FOUND => NoAnswers
```
Comment (jackheslop96) [inline app/models/UserAnswersResponse.scala:52]: API doesn't return a 204, only a 404

File: app/models/UserAnswersResponse.scala [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+  case class BadRequest(code: String) extends UserAnswersResponse
+
+  object BadRequest {
+
+    implicit val reads: Reads[BadRequest] = Json.reads[BadRequest]
+  }
+
+  case object NotAcceptable extends UserAnswersResponse
+
+  case class Other(code: Int, message: String) extends UserAnswersResponse
```
Comment (jackheslop96) [inline app/models/UserAnswersResponse.scala:38]: Doesn't look like we ever do anything with an `Other`, so the information gets lost. Might be better just to throw the exception in the http reads

File: app/models/UserAnswersResponse.scala [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package models
+
+import play.api.http.Status.*
+import uk.gov.hmrc.http.{HttpReads, HttpResponse}
+
+trait UserAnswersResponse
```
Comment (jackheslop96) [inline app/models/UserAnswersResponse.scala:22]: trait can be sealed

**Review by jackheslop96 — APPROVED** [review]

**Review by jackheslop96 — DISMISSED** [review]

**Review by JSquires91 — APPROVED** [review]

**Pull request merged**